File: frontend/src/services/adminService.ts
import api from '@/lib/api'
import type { KnowledgeDocument, ConversationLog, DashboardStats, AdminUser, AuthTokens, StudentAccess } from '@/types'

export interface LoginResponse { access_token: string; refresh_token: string; token_type: string; expires_in: number }
export interface UserResponse { id: number; email: string; full_name: string; role: 'admin' | 'user'; is_active: boolean; created_at: string }

export async function login(email: string, password: string): Promise<{ user: AdminUser; tokens: AuthTokens }> {
  const { data } = await api.post<LoginResponse>('/auth/login', { email, password })
  const tokens: AuthTokens = { accessToken: data.access_token, refreshToken: data.refresh_token, expiresIn: data.expires_in }
  const { data: userData } = await api.get<UserResponse>('/auth/me', { headers: { Authorization: `Bearer ${tokens.accessToken}` } })
  return { user: { id: userData.id.toString(), email: userData.email, name: userData.full_name, role: userData.role, createdAt: userData.created_at }, tokens }
}
export async function logout(): Promise<void> { await api.post('/auth/logout') }
export async function getCurrentUser(): Promise<AdminUser | null> {
  try {
    const { data } = await api.get<UserResponse>('/auth/me')
    return { id: data.id.toString(), email: data.email, name: data.full_name, role: data.role, createdAt: data.created_at }
  } catch { return null }
}

export interface BackendDocument { id: number; title: string; filename: string | null; category: string; chunk_count: number; is_active: boolean; created_at: string; updated_at: string }
export async function getDocuments(params?: { page?: number; pageSize?: number; category?: string; isActive?: boolean }): Promise<{ documents: KnowledgeDocument[]; total: number }> {
  const response = await api.get<{ documents: BackendDocument[]; total: number }>('/admin/documents', { params })
  return { documents: response.data.documents.map(d => ({ id: d.id.toString(), title: d.title, filename: d.filename || undefined, category: d.category as KnowledgeDocument['category'], status: d.is_active ? 'indexed' : 'pending', chunkCount: d.chunk_count, uploadedAt: d.created_at, updatedAt: d.updated_at })), total: response.data.total }
}
export async function uploadDocument(file: File, category: KnowledgeDocument['category'], description?: string): Promise<KnowledgeDocument> {
  const formData = new FormData(); formData.append('file', file); formData.append('category', category); if (description) formData.append('title', description)
  const { data } = await api.post<BackendDocument>('/admin/documents/upload', formData, { headers: { 'Content-Type': 'multipart/form-data' } })
  return { id: data.id.toString(), title: data.title, filename: data.filename || undefined, category: data.category as KnowledgeDocument['category'], status: data.is_active ? 'indexed' : 'pending', chunkCount: data.chunk_count, uploadedAt: data.created_at, updatedAt: data.updated_at, description }
}
export async function deleteDocument(id: string): Promise<void> { await api.delete(`/admin/documents/${id}`) }
export async function reindexDocument(id: string): Promise<void> { await api.post(`/admin/documents/${id}/reindex`) }
export async function getDocumentStats() { const { data } = await api.get('/admin/stats'); return { totalDocuments: data.total_documents, activeDocuments: data.active_documents, totalChunks: data.total_chunks } }

export interface BackendConversationLog { id: string; session_id: string; user_query: string; bot_response: string; sources: any[]; confidence: number; is_fallback: boolean; feedback: 'helpful' | 'not_helpful' | null; response_time_ms: number; created_at: string }
export async function getLogs(params?: { page?: number; pageSize?: number; sessionId?: string }): Promise<{ logs: ConversationLog[]; total: number }> {
  const { data } = await api.get<{ logs: BackendConversationLog[]; total: number }>('/admin/conversations', { params: { page: params?.page, page_size: params?.pageSize } })
  return { logs: data.logs.map(l => ({ id: l.id, sessionId: l.session_id, query: l.user_query || '', response: l.bot_response, sources: l.sources || [], confidence: l.confidence, isFallback: l.is_fallback, feedback: l.feedback, responseTime: l.response_time_ms, timestamp: l.created_at, status: l.is_fallback ? 'fallback' : 'answered' })), total: data.total }
}

export interface BackendDashboardStats { total_conversations: number; unique_users: number; avg_response_time_ms: number; avg_confidence_score: number; fallback_rate: number; helpful_rate: number; active_documents: number; total_chunks: number; period_start: string; period_end: string }
export const DASHBOARD_STATS_PERIOD_DAYS = 7
export const DASHBOARD_STATS_QUERY_KEY = ['dashboard-stats', DASHBOARD_STATS_PERIOD_DAYS] as const
export async function getDashboardStats(days = 7): Promise<DashboardStats> {
  const { data } = await api.get<BackendDashboardStats>('/admin/dashboard', { params: { days } })
  return { totalConversations: data.total_conversations, uniqueUsers: data.unique_users, avgResponseTime: data.avg_response_time_ms, avgConfidence: data.avg_confidence_score, fallbackRate: data.fallback_rate, helpfulRate: data.helpful_rate, totalDocuments: data.active_documents, activeDocuments: data.active_documents, totalChunks: data.total_chunks, periodStart: data.period_start, periodEnd: data.period_end }
}

export interface PFEEvaluationStats {
  period_days: number; period_start: string; period_end: string; sessions: number; total_queries: number
  text_queries: number; voice_queries: number; voice_usage_rate: number; text_usage_rate: number
  fallback_rate: number; resolution_rate: number; helpful_rate: number; feedback_count: number
  avg_response_time_ms: number; sus_response_count: number; sus_avg_score: number | null
}
export async function getPFEEvaluationStats(days = 30): Promise<PFEEvaluationStats> {
  const { data } = await api.get<PFEEvaluationStats>('/admin/pfe-evaluation', { params: { days } })
  return data
}

export interface FallbackQuestion { question: string; count: number; last_seen: string }
export async function getFallbackQuestions(days = 30, limit = 10): Promise<FallbackQuestion[]> { const { data } = await api.get<FallbackQuestion[]>('/admin/fallback-questions', { params: { days, limit } }); return data }
export async function getTopQuestions(days = 7, limit = 10): Promise<{ question: string; count: number }[]> { const { data } = await api.get('/admin/top-questions', { params: { days, limit } }); return data }

interface BackendStudentAccess { id: number; full_name: string; masked_identifier: string; is_active: boolean; created_at: string; updated_at: string }
function mapStudentAccess(student: BackendStudentAccess): StudentAccess { return { id: student.id, fullName: student.full_name, maskedIdentifier: student.masked_identifier, isActive: student.is_active, createdAt: student.created_at, updatedAt: student.updated_at } }
export async function getStudentAccessList(): Promise<StudentAccess[]> { const { data } = await api.get<BackendStudentAccess[]>('/admin/students'); return data.map(mapStudentAccess) }
export async function createStudentAccess(fullName: string, studentIdentifier: string): Promise<StudentAccess> { const { data } = await api.post<BackendStudentAccess>('/admin/students', { full_name: fullName, student_identifier: studentIdentifier }); return mapStudentAccess(data) }
export async function setStudentAccessActive(id: number, isActive: boolean): Promise<StudentAccess> { const { data } = await api.patch<BackendStudentAccess>(`/admin/students/${id}`, { is_active: isActive }); return mapStudentAccess(data) }
export async function deleteStudentAccess(id: number): Promise<void> { await api.delete(`/admin/students/${id}`) }

File: backend/app/services/admin_service.py
from datetime import datetime, timedelta
from typing import Any
import json
import logging

from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncSession

from app.repositories.conversation_repository import ConversationRepository
from app.repositories.document_repository import DocumentRepository
from app.models.admin import DashboardStats, ConversationLog, FallbackQuestion

logger = logging.getLogger(__name__)


class AdminService:
    def __init__(self, db: AsyncSession):
        self.db = db
        self.conversation_repo = ConversationRepository(db)
        self.document_repo = DocumentRepository(db)

    async def get_dashboard_stats(self, days: int = 7) -> DashboardStats:
        end_date = datetime.utcnow()
        start_date = end_date - timedelta(days=days)
        stats = await self.conversation_repo.get_stats(start_date, end_date)
        doc_stats = await self.document_repo.get_total_chunks()
        total_messages = stats.get("total_messages", 0)
        fallback_rate = stats.get("fallback_count", 0) / max(total_messages, 1)
        feedback_count = stats.get("helpful_count", 0) + stats.get("not_helpful_count", 0)
        helpful_rate = stats.get("helpful_count", 0) / max(feedback_count, 1)
        return DashboardStats(
            total_conversations=stats.get("total_conversations_all_time", 0), unique_users=stats.get("unique_users", 0),
            avg_response_time_ms=stats.get("avg_response_time_ms", 0.0), avg_confidence_score=stats.get("avg_confidence", 0.0),
            fallback_rate=fallback_rate, helpful_rate=helpful_rate,
            active_documents=await self.document_repo.count(is_active=True), total_chunks=doc_stats,
            period_start=start_date, period_end=end_date,
        )

    async def get_pfe_evaluation_stats(self, days: int = 30) -> dict[str, Any]:
        """Metrics required by the PFE pilot, calculated only from real persisted interactions."""
        end_date = datetime.utcnow()
        start_date = end_date - timedelta(days=days)
        stmt = text("""
            SELECT
                COUNT(*) FILTER (WHERE m.role = 'user') AS total_queries,
                COUNT(*) FILTER (WHERE m.role = 'user' AND m.sources LIKE '%\"interaction_mode\": \"voice\"%') AS voice_queries,
                COUNT(*) FILTER (WHERE m.role = 'user' AND (m.sources IS NULL OR m.sources NOT LIKE '%\"interaction_mode\": \"voice\"%')) AS text_queries,
                COUNT(*) FILTER (WHERE m.role = 'assistant') AS assistant_answers,
                COUNT(*) FILTER (WHERE m.role = 'assistant' AND m.is_fallback = TRUE) AS fallback_answers,
                COUNT(*) FILTER (WHERE m.role = 'assistant' AND m.feedback = 'HELPFUL') AS helpful_answers,
                COUNT(*) FILTER (WHERE m.role = 'assistant' AND m.feedback = 'NOT_HELPFUL') AS not_helpful_answers,
                AVG(m.response_time_ms) FILTER (WHERE m.role = 'assistant' AND m.response_time_ms IS NOT NULL) AS avg_response_time_ms,
                COUNT(DISTINCT m.conversation_id) AS sessions
            FROM messages m
            WHERE m.created_at >= :start_date AND m.created_at <= :end_date
        """)
        row = (await self.db.execute(stmt, {"start_date": start_date, "end_date": end_date})).mappings().one()
        total = int(row["total_queries"] or 0)
        voice = int(row["voice_queries"] or 0)
        text_count = int(row["text_queries"] or 0)
        answers = int(row["assistant_answers"] or 0)
        fallbacks = int(row["fallback_answers"] or 0)
        helpful = int(row["helpful_answers"] or 0)
        not_helpful = int(row["not_helpful_answers"] or 0)
        feedback_total = helpful + not_helpful
        return {
            "period_days": days, "period_start": start_date, "period_end": end_date,
            "sessions": int(row["sessions"] or 0), "total_queries": total,
            "text_queries": text_count, "voice_queries": voice,
            "voice_usage_rate": voice / total if total else 0.0,
            "text_usage_rate": text_count / total if total else 0.0,
            "fallback_rate": fallbacks / answers if answers else 0.0,
            "resolution_rate": (answers - fallbacks) / answers if answers else 0.0,
            "helpful_rate": helpful / feedback_total if feedback_total else 0.0,
            "feedback_count": feedback_total,
            "avg_response_time_ms": float(row["avg_response_time_ms"] or 0.0),
        }

    async def get_conversation_logs(self, page: int = 1, page_size: int = 50, session_id: str | None = None, has_feedback: bool | None = None, is_fallback: bool | None = None) -> tuple[list[ConversationLog], int]:
        offset = (page - 1) * page_size
        stmt = text("""
            SELECT m.id, m.content AS bot_response, m.sources, m.confidence, m.is_fallback,
                   m.feedback, m.response_time_ms, m.created_at, c.session_id,
                   (SELECT um.content FROM messages um WHERE um.conversation_id = m.conversation_id
                    AND um.role = 'user' AND um.id < m.id ORDER BY um.id DESC LIMIT 1) AS user_query
            FROM messages m JOIN conversations c ON m.conversation_id = c.id
            WHERE m.role = 'assistant' ORDER BY m.created_at DESC LIMIT :limit OFFSET :offset
        """)
        messages = await self.conversation_repo.db.execute(stmt, {"limit": page_size, "offset": offset})
        total_result = await self.conversation_repo.db.execute(text("SELECT COUNT(*) FROM messages WHERE role = 'assistant'"))
        logs = [ConversationLog(
            id=str(row["id"]), session_id=row["session_id"], user_query=row["user_query"] or "",
            bot_response=row["bot_response"], sources=json.loads(row["sources"]) if row["sources"] else [],
            confidence=row["confidence"] or 0.0, is_fallback=row["is_fallback"], feedback=row["feedback"],
            response_time_ms=row["response_time_ms"] or 0, created_at=row["created_at"],
        ) for row in messages.mappings()]
        return logs, total_result.scalar() or 0

    async def get_fallback_questions(self, days: int = 30, limit: int = 10) -> list[FallbackQuestion]:
        end_date = datetime.utcnow(); start_date = end_date - timedelta(days=days)
        questions = await self.conversation_repo.get_fallback_questions(start_date, end_date, limit)
        return [FallbackQuestion(question=q["question"], count=q["count"], last_seen=end_date) for q in questions]

    async def get_top_questions(self, days: int = 7, limit: int = 10) -> list[dict[str, Any]]:
        end_date = datetime.utcnow(); start_date = end_date - timedelta(days=days)
        stmt = text("""SELECT content, COUNT(*) AS count FROM messages WHERE role = 'user'
                     AND created_at >= :start_date AND created_at <= :end_date
                     GROUP BY content ORDER BY count DESC LIMIT :limit""")
        result = await self.conversation_repo.db.execute(stmt, {"start_date": start_date, "end_date": end_date, "limit": limit})
        return [{"question": row["content"], "count": row["count"]} for row in result.mappings()]